### PR TITLE
runner: catch unicode error

### PR DESCRIPTION
**Pull request by damianloew**
## Description

- add handling UnicodeDecodeError exception
- trunner: change codec from utf-8 to ascii
- trunner: combine two try blocks into one in test execution

## Motivation and Context

For the problem known from the following issue: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/295 instead of Unicode Decode Error and aborting test runner, the output is printed.

Before change (test-helloworld prints here the following file content: ab�c):
Screenshot from 2022-02-14 10-47-06

After change:
Screenshot from 2022-03-01 14-55-35

https://github.com/phoenix-rtos/phoenix-rtos-project/issues/175

JIRA: PD-238

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (ia32-generic).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Comment by damianloew** [comment]
I've also tested a solution with `proc.read()`, which could get a whole output at once, but it doesn't work (encounters timeout error, because it can't find eof). That's why there is reading 1 byte each.

**Comment by xvuko** [comment]
Should we use UTF-8 to decode process output? Why ASCII codec is not used?

**Comment by damianloew** [comment]
> Should we use UTF-8 to decode process output? Why ASCII codec is not used?

That's right, ASCII codec could be better. I will change it in next commit here.

File: trunner/testcase.py [hunk 1]
```diff
@@ -171,6 +188,11 @@ def handle(self, proc):
             msg = 'EXCEPTION ' + ('EOF' if isinstance(exc, EOF) else 'TIMEOUT') + '\n'
             self.exception += Color.colorify(msg, Color.BOLD)
             self.handle_pyexpect_error(proc, exc)
+        except UnicodeDecodeError:
+            msg = 'Unicode Decode Error detected!\n'
+            self.exception = Color.colorify(msg, Color.BOLD)
+            self.handle_unicode_error(proc)
+            return
```
Comment (xvuko) [inline trunner/testcase.py:195]: I see a lot of code duplication in exception handling here. Those exact 4 lines appear 3 times. Could this try block be extended to unify catching errors from both harness and exec test?

File: trunner/testcase.py [hunk 2]
```diff
@@ -126,6 +136,13 @@ def handle_pyexpect_error(self, proc, exc):
         self.exception += Color.colorify('GOT:\n', Color.BOLD)
         self.exception += f'{got}\n'
 
+    def handle_unicode_error(self, proc):
```
Comment (xvuko) [inline trunner/testcase.py:139]: I see no separation between private and public class members here. Just pointing that as I hardly see any underscore prefixed functions in trunner so this is out of scope of this review.

File: trunner/testcase.py (hunk 1 above)
Comment (damianloew) [reply trunner/testcase.py:183]: Yeah, I've checked this and it can be done in one try block here. I would like to leave it as it is in handle function.

File: trunner/testcase.py (hunk 2 above)
Comment (damianloew) [reply trunner/testcase.py:127]: I've created issue for this: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/331. I will think about it in meantime.

File: trunner/testcase.py [hunk 3]
```diff
@@ -107,6 +112,11 @@ def exec_test(self, proc):
                 self.exception = Color.colorify(msg, Color.BOLD)
                 self.handle_pyexpect_error(proc, exc)
                 return
+            except UnicodeDecodeError:
+                msg = 'Unicode Decode Error detected!\n'
+                self.exception = Color.colorify(msg, Color.BOLD)
```
Comment (xvuko) [inline trunner/testcase.py]: Can be moved into handle unicode error.  Additionaly passing original exception into handler would allow to print original message that contains information about problematic character.

**Review by xvuko — CHANGES_REQUESTED** [review]

File: trunner/testcase.py [hunk 4]
```diff
@@ -90,23 +90,21 @@ def exec_test(self, proc):
         try:
             # Wait for a prompt
             proc.expect_exact('(psh)% ')
+            if self.exec_cmd:
+                if self.use_sysexec:
+                    self.sysexec(proc)
+                else:
+                    self.exec(proc)
         except (TIMEOUT, EOF) as exc:
             msg = 'Waiting for psh prompt failed!\n'
```
Comment (jsarzy) [inline trunner/testcase.py:99]: Now if execution fails we will get misleading error message. Change it to "exec failed" or similar.

Comment (damianloew) [reply trunner/testcase.py:99]: Thanks, I've missed it.

**Review by xvuko — APPROVED** [review]
